synthesis.go

```diff
@@ -0,0 +1,262 @@
+	CREATE TABLE weights (
+		pos INTEGER REFERENCES seq(pos),
+		codon TEXT NOT NULL REFERENCES codon(codon),
+		weight INTEGER
+	);
+
+	CREATE TABLE codonbias (
+		fromcodon TEXT REFERENCES codon(codon),
+		tocodon TEXT REFERENCES codon(codon),
+		gcbias TEXT
+	);
+
+	CREATE TABLE suggestedfix (
+		id INTEGER PRIMARY KEY AUTOINCREMENT,
+		step INTEGER,
+		start INTEGER REFERENCES seq(pos),
+		end INTEGER REFERENCES seq(pos),
+		gcbias TEXT,
+		quantityfixes INTEGER,
+		suggestiontype TEXT
+	);
+`
+	db.MustExec(createMemoryDbSql)
+	// Insert codons
+	weightTable := make(map[string]int)
+	codonInsert := `INSERT INTO codon(codon, aa) VALUES (?, ?)`
+	for _, aminoAcid := range codontable.AminoAcids {
+		for _, codon := range aminoAcid.Codons {
+			db.MustExec(codonInsert, codon.Triplet, aminoAcid.Letter)
+			weightTable[codon.Triplet] = codon.Weight
+
+			codonBias := strings.Count(codon.Triplet, "G") + strings.Count(codon.Triplet, "C")
+			for _, toCodon := range aminoAcid.Codons {
+				if codon.Triplet != toCodon.Triplet {
+					toCodonBias := strings.Count(toCodon.Triplet, "G") + strings.Count(toCodon.Triplet, "C")
+					switch {
+					case codonBias == toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "NA")
+					case codonBias > toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "AT")
+					case codonBias < toCodonBias:
+						db.MustExec(`INSERT INTO codonbias(fromcodon, tocodon, gcbias) VALUES (?, ?, ?)`, codon.Triplet, toCodon.Triplet, "GC")
+					}
+				}
+			}
+		}
+	}
+
+	// Insert seq and history
+	pos := 0
+	for i := 0; i < len(sequence)-3; i = i + 3 {
+		codon := sequence[i : i+3]
+		db.MustExec(`INSERT INTO seq(pos) VALUES (?)`, pos)
+		db.MustExec(`INSERT INTO history(pos, codon, step) VALUES (?, ?, 0)`, pos, codon)
+		db.MustExec(`INSERT INTO weights(pos, codon, weight) VALUES (?,?,?)`, pos, codon, weightTable[codon])
+		pos++
+	}
+
+	var err error
+	// For a maximum of 100 iterations, see if we can do better
+	for i := 1; i < 100; i++ {
+		suggestions := findProblems(sequence, problematicSequenceFuncs)
+		// If there are no suggestions, break the iteration!
+		if len(suggestions) == 0 {
+			break
+		}
+		for _, suggestion := range suggestions { // if you want to add overlaps, add suggestionIndex
+			// First, let's insert the suggestions that we found using our problematicSequenceFuncs
+			_, err = db.Exec(`INSERT INTO suggestedfix(step, start, end, gcbias, quantityfixes, suggestiontype) VALUES (?, ?, ?, ?, ?, ?)`, i, suggestion.Start, suggestion.End, suggestion.Bias, suggestion.QuantityFixes, suggestion.SuggestionType)
+			if err != nil {
+				return sequence, err
+			}
+		}
+
+		// The following statements are the magic sauce that makes this all worthwhile.
+		// Parameters: step, gcbias, start, end, quantityfix
```

Comment:
nice. Can you explain a bit more why this is a good idea/likely to generate good results?

Reply:
I'll write more in the actual request, but the basic idea is that I am using SQL as a declarative language to choose fixes to the DNA that are most likely to work in a biological context while avoiding infinite recursion. I've attempted a version that doesn't use a declarative language, but it gets REAL hard to read really quick.

Reply:
hm ok, will have to read your detailed explanation to understand.